docs/CONNECTION-FILTERS.md

```diff
@@ -0,0 +1,127 @@
+# curl connection filters
+
+Connection filters is a design in the internals of curl, not visible in its public API. They were added 
+in curl v7.xx.x. This document describes the concepts, its high level implementation and the motivations.
+
+## Filters
+
+A "connection filter" is a piece of code that is responsible for handling a range of operations
+in regard to curl's connections: reading, writing, waiting on external events, connecting and closing down - to name the most important ones.
+
+The most important feat of connection filters is that they can bet stacked on top of each other (or "chained" if you prefer that metaphor). In the very common scenario that you want to retrieve a `https:` url with curl, you need 2 basic things to send the request and get the response: a TCP connection, represented by a `socket` and a SSL instance en- and decrypt over that socket. You write your request to the SSL instance, which encrypts and writes that data to the socket, which then sends the bytes over the network.
+
+With connection filters, curl's internal setup will look something like this (cf for connection filter):
+
+```
+Curl_easy *data         connectdata *conn        cf-ssl        cf-socket
++----------------+      +-----------------+      +-------+     +--------+
+|https://curl.se/|----> | properties      |----> | keys  |---> | socket |--> OS --> network
++----------------+      +-----------------+      +-------+     +--------+
+
+ Curl_write(data, buffer)
+  --> Curl_cfilter_write(data, data->conn, buffer)
+       ---> conn->filter->write(conn->filter, data, buffer)
+```
+
+While connection filters all do different things, they look the same from the "outside". The code in `data` and `conn` does not really know **which** filters are installed. `conn` just writes into the first filter, whatever that is.
+
+Same is true for filters. Each filter has a pointer to the `next` filter. When SSL has encrypted the data, it does not write to a socket, it writes to the next filter. If that is indeed a socket, or a file, or a HTTP/2 connection is of no concern to the SSL filter.
```

Comment:
each 'filter' itself maybe a lego block though have you done analysis on impact on streamability eg. will we introduce logical filters that branch ?

Reply:
I don't think we need to speculate too much about what we *might* do or not do in the future. This work is a refactor and remodel of how we deal with protocols and "layers" today. If we need to make it more fancy in the future, then let's do that in the future.

Reply:
I am not able to prohibit or demand future changes to curl. I can (and will) not write a rule book here.

With the goal of Proxying over HTTP/2, the current implementation needs to change. The current way to "stacking" `SSL*`either on a socket or on another `SSL*` to not work for this. curl needs to be able to become more flexible in how connections are setup and handled. 

If we do a `https:` transfer over a HTTP/2 proxy tunnel, we need a SSL, talking to a HTTP/2 session, talking to an SSL instance, talking to a socket. This is a way to model such cases, so that we can combine these thing in the manner we need.